Metal: Use setBytes:length:atIndex for arguments

shoaibkamil opened the pull request:
Instead of allocating a temporary `MTLBuffer`, if the arguments
size is less than 4KB, use `setBytes:length:atIndex`.  
See https://developer.apple.com/reference/metal/mtlcomputecommandencoder/1443159-setbytes?language=objc#discussion for more information.

This change should never be slower than the existing approach, and helps remove one source of overhead for small pipelines.

CC @estollnitz

abadams commented on src/runtime/metal.cpp:
```diff
@@ -671,14 +679,29 @@ WEAK int halide_metal_run(void *user_context,
+        char *args_ptr;
+
+        if (total_args_size < 4096) {
+            small_args_buffer = (uint8_t*)malloc(sizeof(uint8_t) * total_args_size);
```
Is it worth just using 4K of stack here?

shoaibkamil replied:
Probably; I'll make that change.

zvookin commented:
The docs seem to indicate "setBytes:length:atIndex:" is iOS 8.3 and later, not iOS 8.0 and later which is the baseline for Metal. Idea was to run on anything that supports Metal unless there's a strong reason not to. Anyone more in the know able to comment on this? If it is indeed not supported and offers performance wins of note on new enough versions of Metal, we cna either case on the Metal version or check if the selector is supported.

shoaibkamil commented:
@zvookin You're right, this appears to be iOS 8.3+.  I can implement either a selector check or version check; please let me know which you prefer.

Re: performance wins, we have an internal version of the Metal runtime with several improvements for running multiple small pipelines in a row.  This change gives us ~30% performance improvement, though the improvement is less without our other changes, for which I'll open separate PRs.

zvookin commented:
I generally prefer selector checks because they tie the choice of which
code to use to a specific property of the API. The downside is as there
gets to be many cases, each one requires a selector test where the version
is only requested once. Though that gets into whether the decision is
cached somehow.  I expect you or someone working on this has more expertise
than I do.

Very glad to see this backend getting further improvements!

-Z-

On Sat, Mar 25, 2017 at 8:35 AM Shoaib Kamil <notifications@github.com>
wrote:

> @zvookin <https://github.com/zvookin> You're right, this appears to be
> iOS 8.3+. I can implement either a selector check or version check; please
> let me know which you prefer.
>
> Re: performance wins, we have an internal version of the Metal runtime
> with several improvements for running multiple small pipelines in a row.
> This change gives us ~30% performance improvement, though the improvement
> is less without our other changes, for which I'll open separate PRs.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/halide/Halide/pull/1944#issuecomment-289219184>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABbqFFfVUCHCHbrbn9KzOevghsDeyiKlks5rpTQ0gaJpZM4Moz8c>
> .
>

shoaibkamil commented:
Sorry for the delay on this.  I've rebased and added a selector check that is cached (and invalidated if the device pointer changes).  I've also addressed @abadams's comment and made the temporary buffer live on the stack.

zvookin commented on src/runtime/metal.cpp:
```diff
@@ -195,6 +210,11 @@ struct module_state {
 
+// API Capabilities.  If more capabilities need to be checked,
+// this can be refactored to something more robust/general.
+bool halide_metal_api_supports_set_bytes;
```
These need to be WEAK and should not be prefixed with "halide_" as they are not exported. I'm guessing that if they are not WEAK, the runtime linker will insist that they are prefixed with "halide_". Is there a desire to export the variables or did you run into an issue with this? I can just merge it and make the fix afterward if that is expedient, but figured I'd check in re: intent and history first.

shoaibkamil replied:
No, you're right-- they should be WEAK, and they should not be exported.  I'll change this & push shortly.

zvookin commented:
Apologies for the delay. Was surprised to find this had not been merged yet.
